src/types/CustomLoader.ts

```diff
@@ -0,0 +1,16 @@
+import * as Yup from "yup";
+
+export type CustomLoader = {
+  url: string;
+  text: string;
+  title: string;
+};
+
+export const CustomLoaderSchema = Yup.object().shape<CustomLoader>({
+  url: Yup.string().required(),
+  text: Yup.string().required(),
+  title: Yup.string().required(),
+});
+
+export const isCustomLoader = (data: {}): data is CustomLoader =>
+  CustomLoaderSchema.isValidSync(data);
```

Comment:
[note] This implements a pattern for validating arbitrary data (eg. an object that is read out of our firestore database) and being able to properly validate that it conforms to a type, then 'narrow it's type' correctly.

This is basically the 'correct way' (at least one of them) for us to avoid arbitrarily asserting unknown data to a type with `as Foo` and similar hacky patterns.

eg. This could be used with `.filter(isCustomLoader)` and similar to achieve this.

We could also use this pattern to check data (eg. venues) when we read them out of the firestore database, to ensure they still match our expected/desired types. This would allow us to 'fail fast' and catch weird edge cases/bugs even when someone has 'arbitrarily hacked the data in the database' to include invalid formats/etc.

**Further Reading:**

- https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates

(cc // @mike-lvov @denisdimitrov @sparkletown/engineering FYI)

Reply:
(backlogged to document properly in https://github.com/sparkletown/internal-sparkle-issues/issues/569)